### ISSUE TITLE
2-Factor Authentication Support

### ISSUE THREAD

**Issue by colemantrenton** [bluesky-social/social-app#1071]
**Is your feature request related to a problem? Please describe.**

2Factor authentication is generally a very handy feature for security purposes as passwords and logins sometimes fail. 2Factor Authentication has become a very common addition to the login suite for protection.

**Describe the solution you'd like**

Implementation and support of some kind of 2FA support for accounts

**Describe alternatives you've considered*

Didn't have anything in mind particularly, just thinking of security lock downs on user accounts

**Comment by Yuuyatails** [comment]
I'm no programmer, but I can say that 2FA is needed. The problem is what kind of 2FA will be implement. There four types of 2FAs: email-based, SMS-based, authenticator-based and FIDO-key-based.

Out of all four 2FA methods, SMS-based is the least secure method because it is known to be vulnerable to SIM Swap Attacks, and operating the SMS-based 2FA can be costly.

My suggestion would be implement only the email-based, authenticator-based and FIDO-key-based methods.

**Comment by Zero3K** [comment]
Any news regarding this?

**Comment by BirdInFire** [comment]
Yeah make this a priority please

**Comment by deviantsemicolon** [comment]
This needs to happen ASAP. We can't have a social network without 2FA. Put this before federation.

**Comment by Eutropios** [comment]
> This needs to happen ASAP. We can't have a social network without 2FA. Put this before federation.
Strong agree, especially in a time where governments are scrutinizing the security of social media platforms.

**Comment by surfdude29** [comment]
OAuth needs to be implemented before 2FA can be added. In [this post](https://bsky.app/profile/dholms.xyz/post/3kkqzvbbza22v) @/dholms said:

> "OAuth is coming & will give us some good options for implementing 2fa 👌"

The Bluesky team are in the process of implementing OAuth, especially @/matthieusieben [in this PR](https://github.com/bluesky-social/atproto/pull/1958).

**Comment by BirdInFire** [comment]
> OAuth needs to be implemented before 2FA can be added. In [this post](https://bsky.app/profile/dholms.xyz/post/3kkqzvbbza22v) @/dholms said:
> 
> > "OAuth is coming & will give us some good options for implementing 2fa 👌"
> 
> The Bluesky team are in the process of implementing OAuth, especially @/matthieusieben [in this PR](https://github.com/bluesky-social/atproto/pull/1958).

Please warn here once is live (because I recently successfully removed all services that do not have 2FA from the one I use, so for me it's a deal breaker (for now) :)
Will join once done.

**Comment by Eutropios** [comment]
> OAuth needs to be implemented before 2FA can be added. In [this post](https://bsky.app/profile/dholms.xyz/post/3kkqzvbbza22v) @/dholms said:
> 
> > "OAuth is coming & will give us some good options for implementing 2fa 👌"
> 
> The Bluesky team are in the process of implementing OAuth, especially @/matthieusieben [in this PR](https://github.com/bluesky-social/atproto/pull/1958).

@surfdude29 I appreciate the update! I don't mean to come off as pushy. I am greatly appreciative of the work you all do

**Comment by surfdude29** [comment]
> @surfdude29 I appreciate the update! I don't mean to come off as pushy. I am greatly appreciative of the work you all do

@Eutropios No worries! Fwiw your comment didn't come off as pushy to me, and I agree it's very important for Bluesky to implement 2FA as soon as they can.

Also, just to clarify, I make small contributions every now and then, but I'm not part of the Bluesky team or anything, I just thought it might be helpful to pass on here the latest that I'd heard about 2FA :)

**Comment by KenGoBlue** [comment]
> @Yuuyatails My suggestion would be implement only the email-based, authenticator-based and FIDO-key-based methods.

I signed on just to second this — especially for authenticator-based/OTP 2FA.

**Comment by surfdude29** [comment]
Quick update for those following, @/pfrazee [posted](https://bsky.app/profile/pfrazee.com/post/3kqsmtccfq22u) just over an hour ago:

> An email-based 2FA will be in the next release, and more factors will come in the next few months after our OAuth layer lands

The email auth factor landed in #3602 and the relevant atproto PRs look to be https://github.com/bluesky-social/atproto/pull/2416 and https://github.com/bluesky-social/atproto/pull/2419.

**Comment by surfdude29** [comment]
Another quick update, 2FA using email is now available on the web 🎉

I tried it and it worked great 👌

Just something to be aware of though, if you enable it now on the web, it won't be possible for you to log in using the iOS or Android app until you have the 1.79 update (unless you disable 2FA on the web first, of course).

The 1.79 update is currently awaiting review by the app stores and will hopefully be available in the next few days.

**Comment by surfdude29** [comment]
The 1.79 update is now rolling out in the app stores with 2FA via email 🚀

(last update from me, sorry for sending so many)

**Comment by BirdInFire** [comment]
> The 1.79 update is now rolling out in the app stores with 2FA via email 🚀
> 
> (last update from me, sorry for sending so many)

Thanks for the head up, but it will be without me on email as not secure, hope you will not do the same mistake ok more secure one (webauthn/passkey to "hide" it behind otp)

But congratulations for this first step !

**Comment by surfdude29** [comment]
> > The 1.79 update is now rolling out in the app stores with 2FA via email 🚀
> > (last update from me, sorry for sending so many)
> 
> As simple head up any news over the SSO (for more advanced 2FA ?)

@/bnewbold [said on 4 June about OAuth](https://github.com/bluesky-social/atproto/discussions/2544#discussioncomment-9657197):

> We don't commit to specific timelines publicly, but we have done a lot of work on it and it is finally coming together, will hopefully be rolling out in stages in the next few weeks.

@/matthieusieben is still hard at work, putting what are hopefully finishing touches on the OAuth implementation – the latest PR is here: https://github.com/bluesky-social/atproto/pull/2483

**Comment by meguroyama** [comment]
I'd like to ask if OATH-TOTP or passkey (webauthn) support is being considered?

**Comment by scottdotweb** [comment]
Hello, just wondering if there's any news on TOTP support coming as there's not been a team update on this issue since June. Thanks!

**Comment by mp-pinheiro** [comment]
Adding my support for this as well. With the platform growing at such rapid rates, it's just a matter of time for us to start hearing the horror stories of lost accounts and identity scams. This is not only a crucial security feature by itself, it's also a very clear money save for Bsky in support tickets for these kinds of issues. With robust 2FA, it's guaranteed to reduce it tenfold.

Also, mentioning [this issue](https://github.com/bluesky-social/pds/issues/99), as it seems for self-hosted PDS's, not even email 2FA is supported, which is very concerning.

**Comment by Leseratte10** [comment]
>  it's also a very clear money save for Bsky in support tickets for these kinds of issues. With robust 2FA, it's guaranteed to reduce it tenfold.

I doubt that. Way too many people are too stupid to know that "You have to backup your recovery keys" means that they have to really, actually backup their recovery keys, and then they'll come whine to the support when they lose their only registered Authenticator and can't access their account anymore. 

When they do add proper 2FA, I hope that A) they make it crystal clear that you need to backup your 2FA restore keys (and IMO they should enforce that by asking you for one of the recovery keys like a week after you enabled 2FA to make sure you actually have these stored), and B) they need to implement it like Gitlab who [refuses to have support reset the 2FA data for any reason](https://about.gitlab.com/blog/2020/08/04/gitlab-support-no-longer-processing-mfa-resets-for-free-users/) even if it means the user will lose his account. 

If all you need to do is send a message to the support to get 2FA disabled again, then you might as well skip it, because it would yet again mean that control over the email account alone would also mean control over the Bluesky account. An attacker who gets access to the email account but not the 2FA authenticator could then just contact Bluesky to reset it, which voids the entire purpose of 2FA.

**Comment by AndyNortrup** [comment]
> I get what you're saying, and I agree to some extent, but in my experience the great majority of attacks like these stop at the "please put in your 2FA codes". As much as people are prone to not backing up recovery codes, they are even more prone to using insecure, leaked passwords for both their accounts and emails, which a proper 2FA flow stops (especially if well enforced).

If you want serious people to use BlueSky you had better have more robust 2FA. There are numerous examples in lots of platforms particularly X where significant accounts and businesses were compromised by sim swap attacks and email takeovers (sometimes backed by a sim-swap).  Yes there will be users who lose their backup codes. But if you want government accounts, companies, and individual users with large followings to know that they aren't going to have some yahoo take over their account and start spewing crypto spam, abuse material, or in the case of companies actually try to move financial markets.

I would like to see support for:
- Passkeys
- Yubikeys
- TOTP codes that I can have in my authenticator app of choice.

**Comment by mp-pinheiro** [comment]
> I would like to see support for:
> 
> * Passkeys
> * Yubikeys
> * TOTP codes that I can have in my authenticator app of choice.

I agree completely. My response was to address the reply that 2FA not being useful if there's no enforcement when people can email the company and just deactivated, which there is validity to that claim, but IMO we shouldn't give up on 2FA because of it. Quite the contrary, there must be a firm stance on accounts that lost keys will not recover access and that's just that.

In fact, Bsky is already very clear in if you lose your account doing say, a badly executed PDS Account Migration, that it leads to loss of data / access and they can do nothing about it. So it seems like this is already their stance.

+1 to all the features you listed.

**Comment by sydseter** [comment]
Please vote if you would like to see support for:

Passkeys
Yubikeys
TOTP codes that I can have in my authenticator app of choice.

**Comment by samleibowitz** [comment]
👍 to MFA generally, and Authenticator apps of choice more specifically.  This is the kind of thing that heads off problems in the long run, and with the recent explosion in the user base i suspect that soon, we will either be glad it exists or very much regret that it doesn’t.

**Comment by melizeche** [comment]
+1 on

TOTP
Passkeys
Yubikeys

**Comment by schuyler** [comment]
Strong +1 for TOTP

**Comment by meguroyama** [comment]
As Bluesky’s user base grows, robust authentication like passkeys and OATH-TOTP become crucial. Email verification is a good start, but isn’t foolproof—emails can be compromised, leading to breaches. Passkeys offer a more secure, user-friendly authentication, reducing weak or reused passwords. However, these methods can be complex for some users and may create entry barriers. Despite that, enhancing security is very important for protecting data and trust as Bluesky expands. I ask that we balance usability with security to ensure ongoing success in an increasingly targeted digital space.

**Comment by pedrostephano** [comment]
Yes please. And echoing a previous comment, please avoid SMS based 2FA. Not secure enough in the presence of SIM swap fraud.

**Comment by sydseter** [comment]
There is also another issue specifically for passkeys: https://github.com/bluesky-social/social-app/issues/1164

Please vote 👍

**Comment by VirtueAvatar** [comment]
It's particularly concerning that this issue has been marked as "on the roadmap" back in August 2023. 

How is decent MFA not considered one of the highest priorities?

**Comment by barrywimlett** [comment]
Authenticator and FIDO/hardware support please :D

**Comment by lbmoore** [comment]
Passkeys are needed ASAP.

**Comment by diogo-fernan** [comment]
+1 on MFA options.

**Comment by MarkBennett** [comment]
I recognise this is another comment, but in the interest of reducing spam for maintainers and others watching this issue, can people please do a 👍🏼 reaction on the issue itself, rather than leaving new comments supporting this feature?

![Screenshot 2024-12-05 at 11 05 44 AM](https://github.com/user-attachments/assets/d6cb0ddc-7094-4e47-9d18-6d7ea8510166)

This won't spam notifications out to everyone, and will also increase the rank of this issue in the tracker.

There are also multiple comments supporting passkeys or against SMS, consider adding a 👍🏼 reaction to one of those, rather than leaving a new comment which again spams others.

Thank-you and sorry for being the grouchy old man in this issue! Love to all! ❤️

**Comment by MarkBennett** [comment]
> My prior comment was intended to provide a rough order of priority to MFA implementations. I see that I wasn't clear about that. My apologies for that.
> 
> Please refrain from calling user feedback "SPAM". It's quite off putting.

Sorry, wasn't trying to put you off @ericjmorey. Just making a general comment and wasn't thinking of your comment in particular. All good.

**Comment by mp-pinheiro** [comment]
> I recognise this is another comment, but in the interest of reducing spam for maintainers and others watching this issue, can people please do a 👍🏼 reaction on the issue itself, rather than leaving new comments supporting this feature?
> 
> ![Screenshot 2024-12-05 at 11 05 44 AM](https://private-user-images.githubusercontent.com/61872/392938595-d6cb0ddc-7094-4e47-9d18-6d7ea8510166.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM0Mzc3OTUsIm5iZiI6MTczMzQzNzQ5NSwicGF0aCI6Ii82MTg3Mi8zOTI5Mzg1OTUtZDZjYjBkZGMtNzA5NC00ZTQ3LTlkMTgtNmQ3ZWE4NTEwMTY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA1VDIyMjQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4MDU4YzRmYTU1MDIxZGNjMTVhNjM4MzE0NjRhYmVkNjdkYjU4NjA5MDZjYmExOTgxNjE3NmFiMDdiOGRlMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AqmFmD8ohQA9WIuVDHuwBU5WZfouXE9GjswI5dXYbL0)
> 
> This won't spam notifications out to everyone, and will also increase the rank of this issue in the tracker.
> 
> There are also multiple comments supporting passkeys or against SMS, consider adding a 👍🏼 reaction to one of those, rather than leaving a new comment which again spams others.
> 
> Thank-you and sorry for being the grouchy old man in this issue! Love to all! ❤️

Also, while I agree with the sentiment, people voicing their support in new comments does a lot to emphasize how wanted a feature is, especially when people (like in this case) share their experiences with the different types of MFA they engage with. You can always unsubscribe to the thread to stop being notified. I think it's important that the devs are also notified and see how desired certain issues are. As a repo maintainer myself, marking reactions is good, but it's when people open and reply to issues that I usually see them.

**Comment by lexmechanicHUB** [comment]
This is absolutely necessary, as the importance of bluesky increases, so will the incentives for bad actors to compromise accounts.

**Comment by SlyAceZeta** [comment]
> I recognise this is another comment, but in the interest of reducing spam for maintainers and others watching this issue, can people please do a 👍🏼 reaction on the issue itself, rather than leaving new comments supporting this feature?
> 
> This won't spam notifications out to everyone, and will also increase the rank of this issue in the tracker.
> 
> There are also multiple comments supporting passkeys or against SMS, consider adding a 👍🏼 reaction to one of those, rather than leaving a new comment which again spams others.
> 
> Thank-you and sorry for being the grouchy old man in this issue! Love to all! ❤️

If anyone is concerned about spam, they can simply customize the notifications they receive:

![image](https://github.com/user-attachments/assets/e5280a33-fe0d-418b-9cb0-676ea4444951)
![image](https://github.com/user-attachments/assets/3af9014a-d160-4ebe-8e7c-4c564e1e3b1c)

**Comment by smurfhunter** [comment]
+1 to this

**Comment by mmraz** [comment]
> There four types of 2FAs: email-based, SMS-based, authenticator-based and FIDO-key-based.

Passkeys are also multi-factor and allow the client side to choose their own implementation.

On my laptop I use a fingerprint reader and a yubikey + PIN as passkey authorizations.   On my phone it's FaceID, Google Authenticator and SMS as a last resort.

**Comment by ChicagoDave** [comment]
I'm trying to login using the API agent and the login page has a 2FA Confirmation field for XXXXX-XXXXX which I received in my email, but the field won't accept the code, complaining it's not in the right format.

**Comment by badja-com** [comment]
It is the end of 2024, nearly 2025. MFA is no longer optional. With the massive increase in users, Blue Sky is a cyber target.

Blue Sky Development - Add MFA ==NOW==.

**Comment by badja-com** [comment]
> I recognise this is another comment, but in the interest of reducing spam for maintainers and others watching this issue, can people please do a 👍🏼 reaction on the issue itself, rather than leaving new comments supporting this feature?
> 
> This won't spam notifications out to everyone, and will also increase the rank of this issue in the tracker.
> 
> There are also multiple comments supporting passkeys or against SMS, consider adding a 👍🏼 reaction to one of those, rather than leaving a new comment which again spams others.
> 
> Thank-you and sorry for being the grouchy old man in this issue! Love to all! ❤️

NO - this is too important and too urgent. Make noise and flare.

**Comment by legowerewolf** [comment]
At the risk of contributing to the annoyance...

> If anyone is concerned about spam, they can simply customize the notifications they receive

There's no way to differentiate between meaningful updates from the team in this issue and us general users who just want a feature. Use the reactions and reduce the noise-to-signal ratio.

All that said: Devs, could we get an update on when this is likely to land? Or a summary of what still needs to happen before it does?

**Comment by scottdotweb** [comment]
> NO - this is too important and too urgent. Make noise and flare.

I strongly agree with the sentiments expressed here by various people that 2FA is overdue, but comments like this one are a good way to get developers to lock an issue against any further replies. So please, no yelling.

**Comment by badja-com** [comment]
I accept your advice.

*Alan McNamara*

I recognise the Guringai (Ku-ring-gai) language group of the Garigal
Aboriginal clan as the Traditional Owners and holders of knowledge of the
place where I live and work.

**Comment by 0xdevalias** [comment]
See also:

- https://github.com/bluesky-social/pds/issues/181
- https://github.com/bluesky-social/social-app/issues/1164
- https://github.com/bluesky-social/social-app/issues/2619
- etc

**Comment by gm112** [comment]
Howdy, just chiming in some two cents on something I noticed today. When 2FA is enabled, it might be a good idea to invalidate all existing sessions to enforce 2FA onto them. As of the time of this comment, when enabling Email 2FA, there's no verification(if your e-mail is already verified) and existing sessions seem to still work even without being authorized with 2FA. The current workflow feels like there could be more done to help give users security, or at least that's just my limited take!

**Comment by ChiefGyk3D** [comment]
There's now multiple issues opened with showing how Email MFA is broken including but not limited to this

The obvious fix, TOTP and/or FIDO/WebAuthn support

https://github.com/bluesky-social/social-app/issues/7618

**Comment by fredgolightly** [comment]
Surely even the negligible cost of a 2fa email is more than the use of a 2fa code? Crazy this is still an issue

**Comment by surfdude29** [comment]
Fwiw, my understanding as to why other 2FA methods haven't been implemented yet is because that's dependent on OAuth and the OAuth rollout is still ongoing. For instance, afaik this part of [the OAuth roadmap](https://github.com/bluesky-social/atproto/discussions/2656) hasn't happened yet:

> Bluesky Social app (web and mobile) gets full support for OAuth, and most users are transitioned to this auth flow for day-to-day use.

Hopefully we'll get updated protocol and product roadmaps sometime soon that will shed more light on when we can expect various features, including more 2FA options.

**Comment by richardr1126** [comment]
It is impossible to log in to my account. I have been contacting support for 2 months and made an [isssue](#7390) here last month.

I wasn't sure if this was just a problem with my custom email domain or not. 🫠 We need a fix.

**Comment by GeorgBNM** [comment]
Same here. Will you ever fix it, Bluesky team?

**Comment by argium** [comment]
@pfrazee do you have any ETA on non-email 2FA such as passkeys and authenticator apps?

**Comment by sydseter** [comment]
Please note that neither

NIST: https://pages.nist.gov/800-63-FAQ/#q-b11

or…

OWASP ASVS: https://github.com/OWASP/ASVS/blob/master/5.0/en/0x15-V6-Authentication.md#v63-general-authentication-security

6.3.6 | Verify that email is not used as either a single-factor or multi-factor authentication mechanism. | 3
-- | -- | --


recommend the use of email as 2FA.

**Comment by hermeschen1116** [comment]
Please support passkey or at least OTP as 2FA. Using email to receive the code is so annoying.

**Comment by noktulo** [comment]
I know that the reason that 2FA isn't implemented is that it's dependent on oAuth. But is there any explanation for why oAuth seems to be a year behind schedule?

**Comment by bnb** [comment]
@cloonix then submit a PR if Claude can do it perfectly :)

**Comment by Damariobros** [comment]
> I know that the reason that 2FA isn't implemented is that it's dependent on oAuth. But is there any explanation for why oAuth seems to be a year behind schedule?

This! Please!

**Comment by devilleweppenaar** [comment]
I think this might soon be a focus.

From the recently published [AT Protocol Roadmap - Spring 2026](https://atproto.com/blog/2026-spring-roadmap#account-experience):

> We also plan to improve account security with additional 2FA methods, to complement the existing email 2FA functionality.

They also say this under [recent progress](https://atproto.com/blog/2026-spring-roadmap#recent-progress):

> OAuth Completeness: permissions and permission sets shipped, giving app devs flexibility in requesting access to user accounts. SDK support, cookbooks, and developer documentation for OAuth got big overhauls, making it significantly easier to integrate into client apps.

I sincerely hope this means we see Passkeys support first. 🤞

**Comment by hermeschen1116** [comment]
Is there any updates on this?

**Comment by ccurzio** [comment]
Folks we really need some sort of 2FA that is better than email or SMS. Passkey/Security Key auth would be nice but even TOTP via authenticator app will do.

**Comment by Damariobros** [comment]
STILL no updates???

We need security keys, passkeys, TOTP, and backup codes!

**Comment by Leseratte10** [comment]
Not sure how this advertisement for your website is helpful here.

**Comment by sentientyeet** [comment]
is this a joke lmao how is a basic security feature not part of the supposed twitter killer

**Comment by redcatjack** [comment]
tfw no basic security features still aren't added after **TWO YEARS.**
